harness/api_checkers.go

```diff
@@ -0,0 +1,482 @@
+func (s *APICheckSuite) TestUnknownFormat() {
+	customFormat := "kiss my shiny metal ..."
+	span := s.tracer.StartSpan("Bender")
+
+	err := span.Tracer().Inject(span.Context(), customFormat, nil)
+	if s.opts.CheckInject {
+		s.Equal(opentracing.ErrUnsupportedFormat, err)
+	} else {
+		s.T().Log("CheckInject capability not set, skipping")
+	}
+	ctx, err := s.tracer.Extract(customFormat, nil)
+	s.Nil(ctx)
+	if s.opts.CheckExtract {
+		s.Equal(opentracing.ErrUnsupportedFormat, err)
+	} else {
+		s.T().Log("CheckExtract capability not set, skipping")
+	}
+}
+
+// ForeignSpanContext satisfies the opentracing.SpanContext interface, but otherwise does nothing.
+type ForeignSpanContext struct{}
+
+// ForeachBaggageItem could call handler for each baggage KV, but does nothing.
+func (f ForeignSpanContext) ForeachBaggageItem(handler func(k, v string) bool) {}
+
+// NotACarrier does not satisfy any of the opentracing carrier interfaces.
+type NotACarrier struct{}
+
+// TestInvalidInject checks if errors are returned when Inject is called with invalid inputs.
+func (s *APICheckSuite) TestInvalidInject() {
+	if !s.opts.CheckInject {
+		s.T().Skip("CheckInject capability not set, skipping")
+	}
+	span := s.tracer.StartSpan("op")
+
+	// binary inject
+	err := span.Tracer().Inject(ForeignSpanContext{}, opentracing.Binary, new(bytes.Buffer))
+	s.Equal(opentracing.ErrInvalidSpanContext, err, "Foreign SpanContext should return invalid error")
+	err = span.Tracer().Inject(span.Context(), opentracing.Binary, NotACarrier{})
+	s.Equal(opentracing.ErrInvalidCarrier, err, "Carrier that's not io.Writer should return error")
+
+	// text inject
+	err = span.Tracer().Inject(ForeignSpanContext{}, opentracing.TextMap, opentracing.TextMapCarrier{})
+	s.Equal(opentracing.ErrInvalidSpanContext, err, "Foreign SpanContext should return invalid error")
+	err = span.Tracer().Inject(span.Context(), opentracing.TextMap, NotACarrier{})
+	s.Equal(opentracing.ErrInvalidCarrier, err, "Carrier that's not TextMapWriter should return error")
+
+	// HTTP inject
+	err = span.Tracer().Inject(ForeignSpanContext{}, opentracing.HTTPHeaders, opentracing.HTTPHeadersCarrier{})
+	s.Equal(opentracing.ErrInvalidSpanContext, err, "Foreign SpanContext should return invalid error")
+	err = span.Tracer().Inject(span.Context(), opentracing.HTTPHeaders, NotACarrier{})
+	s.Equal(opentracing.ErrInvalidCarrier, err, "Carrier that's not TextMapWriter should return error")
+}
+
+// TestInvalidExtract checks if errors are returned when Extract is called with invalid inputs.
+func (s *APICheckSuite) TestInvalidExtract() {
+	if !s.opts.CheckExtract {
+		s.T().Skip("CheckExtract capability not set, skipping")
+	}
+	span := s.tracer.StartSpan("op")
+
+	// binary extract
+	ctx, err := span.Tracer().Extract(opentracing.Binary, NotACarrier{})
+	s.Equal(opentracing.ErrInvalidCarrier, err, "Carrier that's not io.Reader should return error")
+	s.Nil(ctx)
```

Comment:
Interesting — I was basing this off of the Tracer interface description [here](https://github.com/opentracing/opentracing-go/blob/master/tracer.go#L100-L109):

> Return values:
> - A successful Extract returns a SpanContext instance and a nil error
> - If there was simply no SpanContext to extract in `carrier`, Extract() returns (nil, opentracing.ErrSpanContextNotFound)
> - If `format` is unsupported or unrecognized, Extract() returns (nil, opentracing.ErrUnsupportedFormat)
> - If there are more fundamental problems with the `carrier` object, Extract() may return opentracing.ErrInvalidCarrier, opentracing.ErrSpanContextCorrupted, or implementation-specific errors.